liquidctl/driver/msi.py

```diff
@@ -0,0 +1,1468 @@
+        yield from super().probe(handle, **kwargs)
+
+    def connect(self, **kwargs):
+        ret = super().connect(**kwargs)
+        self._data = kwargs.pop(
+            "runtime_storage",
+            RuntimeStorage(
+                key_prefixes=[
+                    f"vid{self.vendor_id:04x}_pid{self.product_id:04x}",
+                    f"serial{self.serial_number}",
+                ]
+            ),
+        )
+        self._feature_data = self.device.get_feature_report(0x52, _MAX_DATA_LENGTH)
+        self._fan_cfg = self.get_fan_config()
+        self._fan_temp_cfg = self.get_fan_temp_config()
+
+        return ret
+
+    def initialize(self, **kwargs):
+        pump_mode = kwargs.pop("pump_mode", "balanced")
+        direction = kwargs.pop("direction", "default")
+        if pump_mode == "balanced":
+            pump_mode = "balance"
+        pump_mode_int = _FanMode[pump_mode].value
+        self._data.store("pump_mode", pump_mode_int)
+        dir_int = 0
+        if direction not in ("default", "top", "bottom", "left", "right", "0", "1", "2", "3"):
+            _LOGGER.warning(
+                "unknown direction value: correct values are 0-3 or top, "
+                "button, left, right, default."
+            )
+        if direction in ("1", "left"):
+            dir_int = 1
+        elif direction in ("2", "bottom"):
+            dir_int = 2
+        elif direction in ("3", "right"):
+            dir_int = 3
+        self._data.store("direction", dir_int)
+        self.set_oled_brightness_and_direction(100, dir_int)
+        if pump_mode_int == _FanMode.GAME.value:
+            self.switch_to_game_mode()
+        elif pump_mode_int == _FanMode.BALANCE.value:
+            self.switch_to_balance_mode()
+        elif pump_mode_int == _FanMode.DEFAULT.value:
+            self.switch_to_default_mode()
+        elif pump_mode_int == _FanMode.SILENT.value:
+            self.switch_to_silent_mode()
+        elif pump_mode_int == _FanMode.SMART.value:
+            self.switch_to_smart_mode()
+
+    def get_status(self, **kwargs):
+        self._write((0x31,))
+        array = self._read()
+        assert array[1] == 0x31, "Unexpected value in response buffer"
+        return [
+            ("Fan 1 speed", array[2] + (array[3] << 8), "rpm"),
+            ("Fan 1 duty", array[0x16] + (array[0x17] << 8), "%"),
+            ("Fan 2 speed", array[4] + (array[5] << 8), "rpm"),
+            ("Fan 2 duty", array[0x18] + (array[0x19] << 8), "%"),
+            ("Fan 3 speed", array[6] + (array[7] << 8), "rpm"),
+            ("Fan 3 duty", array[0x1A] + (array[0x1B] << 8), "%"),
+            ("Water block speed", array[8] + (array[9] << 8), "rpm"),
+            ("Water block duty", array[0x1C] + (array[0x1D] << 8), "%"),
+            ("Pump speed", array[0xA] + (array[0xB] << 8), "rpm"),
+            ("Pump duty", array[0x1E] + (array[0x1F] << 8), "%"),
+            ##    ('Temperature inlet', array[12] + (array[13] << 8), '°C'),
```

Comment:
These were a part of the original contributors implementation:
> > All of the code is based on the decompilation of the .NET code. Most likely it was labeled this way in the code. It may not necessarily be functional for every device.
> > If getting a value for your device is not working you should probably remove it.

My device returns constant values instead of anything meaningful: inlet and outlet are always zeroed out, and the sensor temperatures are constant 125C. My conclusion is that either the temperature sensors need to somehow be activated separately or this is a feature only included in some other models (most likely the latter, since the device does not adjust duties without external temperature communication). The thought is that they could be useful in the future if other models are added / the implementation is refined. I can add a brief comment in relation to this.

Reply:
Thanks, adding a comment for the future sounds good to me since the code is still there.